src/kv/idbstore.rs

```diff
@@ -145,3 +155,214 @@ impl Store for IdbStore {
         })
     }
 }
+
+struct ReadTransaction {
+    tx: IdbTransaction,
+    store: IdbObjectStore,
+}
+
+impl ReadTransaction {
+    fn new(idb: &IdbStore) -> Result<ReadTransaction> {
+        let tx = idb.idb.transaction_with_str(OBJECT_STORE)?;
+        Ok(ReadTransaction {
+            store: tx.object_store(OBJECT_STORE)?,
+            tx: tx,
+        })
+    }
+}
+
+#[async_trait(?Send)]
+impl Read for ReadTransaction {
+    async fn has(&self, key: &str) -> Result<bool> {
+        let request = self.store.count_with_key(&key.into())?;
+        let (sender, receiver) = oneshot::channel::<()>();
+        let callback = Closure::once(move || {
+            if let Err(_) = sender.send(()) {
+                warn!("oneshot send failed");
+            }
+        });
+        request.set_onsuccess(Some(callback.as_ref().unchecked_ref()));
+        request.set_onerror(Some(callback.as_ref().unchecked_ref()));
+        receiver.await?;
+        Ok(match request.result()?.as_f64() {
+            Some(v) if v >= 1.0 => true,
+            _ => false,
+        })
+    }
+
+    async fn get(&self, key: &str) -> Result<Option<Vec<u8>>> {
+        let request = self.store.get(&key.into())?;
+        let (sender, receiver) = oneshot::channel::<()>();
+        let callback = Closure::once(move || {
+            if let Err(_) = sender.send(()) {
+                warn!("oneshot send failed");
+            }
+        });
+        request.set_onsuccess(Some(callback.as_ref().unchecked_ref()));
+        request.set_onerror(Some(callback.as_ref().unchecked_ref()));
+        receiver.await?;
+        Ok(match request.result()? {
+            v if v.is_undefined() => None,
+            v => Some(js_sys::Uint8Array::new(&v).to_vec()),
```

Comment:
Not yet clear about the details here but I'm concerned about extra copies of the data here.

Reply:
Especially if go from idb to to rust and then back to js, I wonder if we can pass along opaque values? Probably not at this layer but worth thinking about in the end